syntax = "proto3";

import "google/protobuf/timestamp.proto";

package farm_ng.core;
option go_package = "github.com/farm-ng/genproto/core";

message ProgramdConfig {
  repeated Program programs = 1;
}

message Program {
  string id = 1;
  string name = 2;
  string description = 3;

  message LaunchPath {
    string path = 1;
    // If set, the path will be resolved relative to the
    // value of this environment variable.
    string root_env_var = 2;
  }

  LaunchPath launch_path = 4;
  repeated string launch_args = 5;
}

message ProgramExecution {
  string id = 1;
  int32 pid = 2;
  int32 exit_code = 3;
  google.protobuf.Timestamp stamp_start = 4;
  google.protobuf.Timestamp stamp_end = 5;
}

message ProgramSupervisorStatus {
  message ProgramRunning {
    ProgramExecution program = 1;
  }

  message ProgramStopped {
    ProgramExecution last_program = 1;
  }

  oneof status {
    ProgramRunning running = 1;
    ProgramStopped stopped = 2;
  }

  repeated Program library = 3;
}

message StartProgramRequest {
  string id = 1;
}

message StopProgramRequest {
  string id = 1;
}
